### DISCUSSION TITLE
how to run services with systemd inside a container?

### DISCUSSION THREAD

**Discussion by Azadzadeh** [containers/buildah#4956] [General]
hi! , i tried fetching a debian image and built a container. i entered the container with a bash shell and tried starting a systemd daemon. 
systemctl was not found, so apt install it. now some commands like systemd-analyze are not found. it shows an error about PID1.
how can we run systemctl and systemd tools inside a buildah container?

**Comment by rhatdan** [comment]
Systemd needs to run as Pid1.  This is really not a common pattern. You can easily run systemd in a Podman container.
What is your goal of running systemd within a buildah container?

**Reply by Azadzadeh** [reply]
for testing a systemd service in a simulated linux environment. docker doesn't support systemd by default but as you mentioned, podman does. I thought it would be also possible to do it with buildah too.
In a buildah container, if I install systemctl, some of its commands work but for example is-active doesn't work. if I install systemd, systemctl becomes unavailable.
So if you confirm, I should consider podman if I want this. Thank you

**Reply by Azadzadeh** [reply]
@rhatdan I tried with podman. i get the same set of errors. When I install systemd with apt, I get this error upon using systemctl:
```
System has not been booted with systemd as init system (PID 1). Can't operate. 
Failed to connect to bus: Host is down
```
When I `apt install systemctl`, some commands like start and stop work but subcommands like is-active doesn't. Other related systemd utilites like journalctl and systemd-analyze are also unavailable.

**Reply by rhatdan** [reply]
What command are you executing to start the systemd based container?

**Reply by Azadzadeh** [reply]
created a simple Containerfile (FROM debian:11), then `podman build .` and `podman run -it id bash`

**Reply by rhatdan** [reply]
```
$ cat /tmp/Containerfile 
from debian:11
run apt -y update
run apt -y install systemd

$ $ podman build -t dd /tmp/
STEP 1/3: FROM debian:11
STEP 2/3: run apt -y update
...

$ podman run -ti localhost/dd /bin/systemd
systemd 247.3-7+deb11u4 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Detected virtualization podman.
Detected architecture x86-64.

Welcome to Debian GNU/Linux 11 (bullseye)!

Set hostname to <5b451ab7c364>.

...
```
Now in another terminal you can exec into the container and execute systemctl.
